examples/mnist.py

```diff
@@ -30,48 +30,52 @@
 # and loading it into numpy arrays. It doesn't involve Lasagne at all.
 
 def load_dataset():
-    # We first define some helper functions for supporting both Python 2 and 3.
+    # We first define a download function, supporting both Python 2 and 3.
     if sys.version_info[0] == 2:
         from urllib import urlretrieve
-        import cPickle as pickle
-
-        def pickle_load(f, encoding):
-            return pickle.load(f)
     else:
         from urllib.request import urlretrieve
-        import pickle
-
-        def pickle_load(f, encoding):
-            return pickle.load(f, encoding=encoding)
 
-    # We'll now download the MNIST dataset if it is not yet available.
-    url = 'http://deeplearning.net/data/mnist/mnist.pkl.gz'
-    filename = 'mnist.pkl.gz'
-    if not os.path.exists(filename):
-        print("Downloading MNIST dataset...")
-        urlretrieve(url, filename)
+    def download(filename, source='http://yann.lecun.com/exdb/mnist/'):
+        print("Downloading %s" % filename)
+        urlretrieve(source + filename, filename)
 
-    # We'll then load and unpickle the file.
+    # We then define functions for loading MNIST images and labels.
+    # For convenience, they also download the requested files if needed.
     import gzip
-    with gzip.open(filename, 'rb') as f:
-        data = pickle_load(f, encoding='latin-1')
-
-    # The MNIST dataset we have here consists of six numpy arrays:
-    # Inputs and targets for the training set, validation set and test set.
-    X_train, y_train = data[0]
-    X_val, y_val = data[1]
-    X_test, y_test = data[2]
-
-    # The inputs come as vectors, we reshape them to monochrome 2D images,
-    # according to the shape convention: (examples, channels, rows, columns)
-    X_train = X_train.reshape((-1, 1, 28, 28))
-    X_val = X_val.reshape((-1, 1, 28, 28))
-    X_test = X_test.reshape((-1, 1, 28, 28))
-
-    # The targets are int64, we cast them to int8 for GPU compatibility.
-    y_train = y_train.astype(np.uint8)
-    y_val = y_val.astype(np.uint8)
-    y_test = y_test.astype(np.uint8)
+
+    def load_mnist_images(filename):
+        if not os.path.exists(filename):
+            download(filename)
+        # Read the inputs in Yann LeCun's binary format.
+        with gzip.open(filename, 'rb') as f:
+            data = np.frombuffer(f.read(), np.uint8, offset=16)
```

Comment:
By the way, if you're wondering, `f.seek(16); np.fromfile(f, np.uint8)` doesn't work. `fromfile` expects a filename or an open file, a file-like object such as a `GzipFile` is not accepted. But it's not a lot of data to copy around anyway.